chain/x/zoracle/internal/keeper/report.go

```diff
@@ -6,6 +6,68 @@ import (
 	"github.com/bandprotocol/d3n/chain/x/zoracle/internal/types"
 )
 
+func (k Keeper) AddReport(
+	ctx sdk.Context, requestID int64, dataSet []types.RawDataReport, validator sdk.ValAddress,
+) sdk.Error {
+	request, err := k.GetRequest(ctx, requestID)
+	if err != nil {
+		return err
+	}
+
+	if request.IsResolved || request.ExpirationHeight < ctx.BlockHeight() {
+		// TODO: fix error later
+		return types.ErrRequestNotFound(types.DefaultCodespace)
+	}
+
+	found := false
+	for _, validValidator := range request.RequestedValidators {
+		if validator.Equals(validValidator) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return types.ErrInvalidValidator(types.DefaultCodespace)
+	}
+
+	isFirstReport := true
+	for _, submittedValidator := range request.ReceivedValidators {
+		if validator.Equals(submittedValidator) {
+			isFirstReport = false
+		}
+	}
+
+	if isFirstReport {
+		if int64(len(dataSet)) != k.GetRawDataRequestCount(ctx, requestID) {
+			// TODO: fix error later
+			return types.ErrRequestNotFound(types.DefaultCodespace)
+		}
+	}
+
+	lastExternalID := dataSet[0].ExternalDataID - 1
+	for _, rawReport := range dataSet {
+		if lastExternalID >= rawReport.ExternalDataID {
```

Comment:
Can we check that all raw requests must be reported for the transaction to be successful?

Reply:
The number of reports is checked  and if no report used same  externalID after check exist RawDataRequest, it guarantee that all raw requests must be reported.